contracts/LibSubmarineSimple.sol

```diff
@@ -0,0 +1,195 @@
+        require(block.number - _commitBlockNumber > commitPeriodLength, "You must wait long enough to allow the committing period to end before revealing");
+        ProvenTransaction memory proven_tx;
+        // Commented out for theoretical gas savings
+        (proven_tx.result, /* index */, proven_tx.nonce, /* gasprice */, /* startgas */, proven_tx.to, proven_tx.value, /* data */, /* v */ , /* r */, /* s */, proven_tx.is_contract_creation ) = txProof(commitBlockHash, _proofBlob);
+
+        // TX_PROOF_RESULT_PRESENT = 1;
+        // TX_PROOF_RESULT_ABSENT = 2;
+        require(proven_tx.result == 1, "The proof is invalid");
+        require(proven_tx.value >= _unlockAmount);
+        require(proven_tx.nonce == 0);
+        require(proven_tx.is_contract_creation == false);
+
+        address submarine = ecrecover(
+            _unlockTXHash,
+            vee,
+            keccak256(abi.encodePacked(commitId, byte(1))),
+            keccak256(abi.encodePacked(commitId, byte(0)))
+        );
+
+        require(keccak256(abi.encodePacked(proven_tx.to)) == keccak256(abi.encodePacked(submarine)), "The proven address should match the revealed address, or the txhash/witness is wrong.");
+        commitData[commitId].revealed = true;
+        commitData[commitId].amountRevealed = _unlockAmount;
+        emit Revealed(commitId, _unlockAmount, _witness, commitBlockHash, submarine);
+    }
+    
+    /**
+     * @notice Function called by the submarine address to unlock the session.
+     * @dev warning this function does NO validation whatsoever. ALL validation is done in the reveal. 
+     * @param _commitId committed data; The commit instance representing the commit/reveal transaction
+     */
+    function unlock(bytes32 _commitId) public payable {
+        // Required to prevent an attack where someone would unlock after an unlock had already happened, and try to overwrite the unlock amount.
+        require(commitData[_commitId].amountUnlocked < msg.value, "You can never unlock less money than you've already unlocked.");
+        commitData[_commitId].amountUnlocked = msg.value;
+        commitData[_commitId].unlocked = true;
+        emit Unlocked(_commitId, msg.value);
+    }
+    
+    /**
+     * @notice Finished function can be called to determine if a submarine send transaction has been successfully completed for a given committed
+     * @param _commitId committed data; The commit instance representing the commit/reveal transaction
+     * @return bool whether the commit has a stored submarine send that has been completed for it (0 for failure / not yet finished, 1 for successful submarine TX)
+     */
+    function finished(bytes32 _commitId) public view returns(bool success) {
+        return commitData[_commitId].unlocked 
+            && commitData[_commitId].revealed 
+            && commitData[_commitId].amountUnlocked >= commitData[_commitId].amountRevealed;
+    }
```

Comment:
I really like what you did to simplify the logic of ensuring the right amount is sent. 

Should there be a getter function which retrieves the final commit value and data?

Reply:
line 32 `mapping(bytes32 => CommitData) public commitData;` should provide that, no? If you know the commit ID, since the mapping is public, you can retrieve the data at any time

Reply:
This is for usability purposes and increases the gas cost but is something which I think all implementations would need to add anyways to use the commits on chain.

```
mapping(bytes32 => CommitData) public commitData;

struct CommitData {
        uint128 amountRevealed;
        uint128 amountUnlocked;
        address sender;
        bytes embeddedDAppData;
}

function getFinal(bytes32 _commitId) public view returns (address sender, uint128 amount, bytes data) {
        require(finished(_commitId));

        CommitData memory sesh = commitData[_commitId];
        sender = sesh.sender;
        amount = sesh.amountRevealed;
        data = sesh.embeddedDAppData;
}
``

Reply:
I don't think all client implementations need to store `address sender,` since they should be able to use `msg.sender` of whoever calls their finalize / whatever function for whatever their use-case may be.

As for the embeddedDAppData, that should be handled by the wrapper functions that clients will write around the reveal function. plus, storing arbitrary bytes is going to make gas costs go up :(

Let me chew on this some more - I'm working on example client implementations right now so that will help me concretize with a few specific examples whether or not this should be exposed from a library, or whether we should force client implementations to handle that kind of state storage themselves.